Firestore/core/src/core/query_listener.cc

```diff
@@ -147,8 +148,10 @@ bool QueryListener::ShouldRaiseInitialEvent(const ViewSnapshot& snapshot,
     return false;
   }
 
-  // Raise data from cache if we have any documents or we are offline
-  return !snapshot.documents().empty() || online_state == OnlineState::Offline;
+  // Raise data from cache if we have any documents, have cached results before,
+  // or we are offline.
+  return !snapshot.documents().empty() || snapshot.has_cached_results() ||
+         online_state == OnlineState::Offline;
```

Comment:
It's perhaps "safer" the way it's written now, but I'm curious why we didn't use:

```cpp
return snapshot.has_cached_results() || online_state == OnlineState::Offline;
```

in other words, if `has_cached_results` is true, it doesn't matter what the size of `snapshot.documents()` is. right?

Reply:
hmm... perhpas there's a scenario where snapshot.documents() is not empty, but snapshot.has_cached_results is false? (e.g. some other query downloaded some documents that overlap with this query?)

Reply:
This is a bit nuanced indeed. `has_cached_results()` returns `true` if there _are_ cached results, whether or not this particular snapshot is being raised from cache. Namely, `has_cached_results()` could be `true` but `fromCache()` could be `false` if the snapshot is not being raised from cache (but it _could_ be).

Also, the scenario where `!documents().empty()` is `true` but `has_cached_results()` is `false` is when the query has never been run before BUT the client performed a write that was written into the local cache, in addition to the scenario you outlined where there is a coincidentally query that was previously run.

Reply:
Thanks for clarifying 👍